internal/metadataproviders/openshift/metadata.go

```diff
@@ -0,0 +1,171 @@
+	OpenShiftClusterVersion(context.Context) (string, error)
+	Infrastructure(context.Context) (*InfrastructureAPIResponse, error)
+}
+
+// NewProvider creates a new metadata provider.
+func NewProvider(address, token string) Provider {
+	return &openshiftProvider{
+		address: address,
+		token:   token,
+		client:  &http.Client{},
+	}
+}
+
+type openshiftProvider struct {
+	client  *http.Client
+	address string
+	token   string
+}
+
+func (o *openshiftProvider) makeOCPRequest(ctx context.Context, endpoint, target string) (*http.Request, error) {
+	addr := fmt.Sprintf("%s/apis/config.openshift.io/v1/%s/%s/status", o.address, endpoint, target)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, addr, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", o.token))
+	return req, nil
+}
+
+// OpenShiftClusterVersion requests the ClusterVersion from the openshift api.
+func (o *openshiftProvider) OpenShiftClusterVersion(ctx context.Context) (string, error) {
+	req, err := o.makeOCPRequest(ctx, "clusterversions", "version")
+	if err != nil {
+		return "", err
+	}
+	resp, err := o.client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	res := ocpClusterVersionAPIResponse{}
+	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
+		return "", err
+	}
+	return res.Status.Desired.Version, nil
+}
+
+// ClusterVersion requests Infrastructure details from the openshift api.
+func (o *openshiftProvider) Infrastructure(ctx context.Context) (*InfrastructureAPIResponse, error) {
+	req, err := o.makeOCPRequest(ctx, "infrastructures", "cluster")
+	if err != nil {
+		return nil, err
+	}
+	resp, err := o.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	res := &InfrastructureAPIResponse{}
+	if err := json.NewDecoder(resp.Body).Decode(res); err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+// K8SClusterVersion requests the ClusterVersion from the kubernetes api.
+func (o *openshiftProvider) K8SClusterVersion(ctx context.Context) (string, error) {
+	addr := fmt.Sprintf("%s/version", o.address)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, addr, nil)
+	if err != nil {
+		return "", err
+	}
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", o.token))
+
+	resp, err := o.client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	res := k8sClusterVersionAPIResponse{}
+	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
+		return "", err
+	}
+	version := res.GitVersion
+	if strings.Contains(version, "+") {
+		version = strings.Split(version, "+")[0]
+	}
+	return version, nil
+}
+
+type ocpClusterVersionAPIResponse struct {
+	Status struct {
+		Desired struct {
+			Version string `json:"version"`
+		} `json:"desired"`
+	} `json:"status"`
+}
+
+// InfrastructureAPIResponse from OpenShift API.
+type InfrastructureAPIResponse struct {
+	Status InfrastructureStatus `json:"status"`
+}
+
+type InfrastructureStatus struct {
```

Comment:
Can you add comments to exported structs and fields?  External references would be helpful if they exist.